### ISSUE TITLE
Feature request: promise chaining for preload calls

### ISSUE THREAD

**Issue by mxmul** [sveltejs/sapper#1335]
**Is your feature request related to a problem? Please describe.**

Data fetched by a `preload()` function in a layout component isn't available in the `preload()` functions of individual pages, which makes it difficult to express some data dependencies. 

For example, say I have a multi-page app: I'd like to fetch some data from my REST api to render the layout component (e.g. user info required to render a header), and some other page-specific data on each page. When using this REST api, I must fetch an access token before calling any other endpoints  – it's awkward to ensure that the access token gets fetched _exactly once_ before any of the other data fetching.

**Describe the solution you'd like**

Ideally the `preload` function for a page would receive an additional argument, providing a reference to the `preload` results from its layout component(s). E.g:

```html
<!-- src/routes/_layout.svelte -->
<script context="module">
  export async function preload(page, session) {
    const token = await this.fetch(`${API}/access_token`);
    const userData = await this.fetch(`${API]/me`, {
      headers: {'Authorization': token }
    });
    return { token, userData };
  }
</script>
```

```html
<!-- src/routes/some-page.svelte -->
<script context="module">
  export async function preload(page, session, layoutProps) {
    const { token } = await layoutProps[0];
    const somePageData = await this.fetch(`${API]/somePage`, {
      headers: {'Authorization': token }
    });
    return { somePageData };
  }
</script>
```

**Describe alternatives you've considered**

The same behavior can be achieved by pulling the shared data fetching into a shared module, and memoizing the result. E.g:

```js
/* src/api.js */
let tokenPromise;

export function fetchAccessToken({ fetch = fetch }) {
  if (tokenPromise == null) {
    tokenPromise = fetch(`${API}/access_token`);
  }
  return tokenPromise;
}
```
 ```html
<!-- src/routes/_layout.svelte -->
<script context="module">
  import { fetchAccessToken } from "../api";

  export async function preload(page, session) {
    const token = await fetchAccessToken({ fetch: this.fetch });
    const userData = await this.fetch(`${API]/me`, {
      headers: {'Authorization': token }
    });
    return { userData };
  }
</script>
```

```html
<!-- src/routes/some-page.svelte -->
<script context="module">
  import { fetchAccessToken } from "../api";

  export async function preload(page, session) {
    const token = await fetchAccessToken({ fetch: this.fetch });
    const somePageData = await this.fetch(`${API]/somePage`, {
      headers: {'Authorization': token }
    });
    return { somePageData };
  }
</script>
```

Getting this right and avoiding race conditions is a bit tricky however.

**How important is this feature to you?**
I think it's a nice-to-have; there are ways to enforce the ordering in application code, but it could be more ergonomic.

**Comment by benmccann** [comment]
This is a really good question.

I don't know if we'd always want to do the fetches sequentially because then we're taking a performance hit in cases where they can be done in parallel.

As a higher-level note, I think this sort of demonstrates an issue with the framework vs. library approach. It's really hard to control stuff like this when the framework is calling your code vs when you're calling a library. We've been talking about modularizing Sapper, so perhaps we should think about what it might look like to make Sapper more of a library.

**Comment by benmccann** [comment]
There's some discussion happening on Discord right now that seems related to what you're trying to accomplish. Perhaps you'd want to consider putting the user access token in the session store

https://discord.com/channels/457912077277855764/473466028106579978/735914412736512050

**Comment by mxmul** [comment]
> I don't know if we'd always want to do the fetches sequentially because then we're taking a performance hit in cases where they can be done in parallel.

There wouldn't have to be a performance hit. The `preload`s would have to be invoked sequentially, but there's nothing stopping multiple `preload`s from fetching data concurrently. In pseudo-code, I'm thinking it could be implemented something like this:

```js
let promises = [];
for (const i = 0; i < page.parts.length; i++) {
  promises.push(
    page.parts[i].preload.call(
      ctx,
      { host, path, query, params },
      session,
      promises.slice(0, i),
    )
  )
}
const preloaded = await Promise.all(promises);
```

In my original example, the page `preload` is awaiting the layout `preload` so the fetches happen sequentially. In any existing Sapper applications, there would be no such `await` and fetches could happen in parallel.

> Perhaps you'd want to consider putting the user access token in the session store

I think you're referring to something [like this](https://sapper.svelte.dev/docs#Seeding_session_data) (populating session from an express middleware)? I think that approach does work, but it's now server-only logic; being able to run the exact same preload code client-side would be more powerful.

**Comment by benmccann** [comment]
> I think you're referring to something like this (populating session from an express middleware)? I think that approach does work, but it's now server-only logic

I haven't used session much, but my understanding is you can update it client-side as well. The docs say: "It is a writable sore, meaning you can update it with new data (for example, after the user logs in) and your app will be refreshed"

> The preloads would have to be invoked sequentially, but there's nothing stopping multiple preloads from fetching data concurrently.

Ah, that's a pretty nice solution

The one thing I still wonder about is what if you have nested layouts. I guess that's what `layoutProps[0]` indicates in your example? I wonder if it's brittle though. If you add or remove a layout above another one then you may need to update many instances of `layoutProps[i]`. It also forces components to be used in a consistent hierarchy. What if you have a single component that is used under one layout in one place and two layouts in another place?

Perhaps `layoutProps` should be a map with some type of key to represent which layout it is? Or maybe all results should be merged into a single object and you need to make your key names unique? (and we provide a warning in the case that they're not)
